feat: song search

tomast1337 opened the pull request.

tomast1337 committed:
feat: implement getFeaturedSongs method in SongService and add corresponding tests

- Added getFeaturedSongs method to SongService to retrieve featured songs across various timespans.
- Implemented tests for getFeaturedSongs to ensure correct functionality and handling of empty results.
- Removed SongBrowser module and related files as part of the refactor to streamline the service structure.
- Introduced new PageDto class for pagination handling in the database module.

tomast1337 committed:
feat: enhance SongController with advanced song retrieval features

- Implemented various query modes in getSongList method to support fetching featured, recent, category-based, and random songs.
- Added error handling for invalid query parameters and improved response types using PageDto.
- Introduced searchSongs method for keyword-based song searches with pagination.
- Updated Swagger documentation to reflect new query parameters and response structures.

tomast1337 committed:
Merge branch 'develop' of github.com:OpenNBS/NoteBlockWorld into feature/song-search

tomast1337 committed:
refactor: remove SongBrowserModule from app.module.ts

Bentroen committed:
Merge branch 'develop' into feature/song-search

tomast1337 committed:
feat: enhance song retrieval with new search and sorting capabilities

- Introduced  for improved query handling, allowing filtering by title, category, uploader, and sorting by various criteria.
- Updated  method in  to support new query parameters and sorting logic.
- Added new endpoints for fetching featured songs and available categories.
- Refactored frontend components to align with updated API endpoints and enhance song search functionality.
- Improved error handling and loading states in the frontend for better user experience.

tomast1337 committed:
feat: add SearchButton component to enhance song search functionality

- Introduced a new SearchButton component for searching songs.
- Integrated the SearchButton into the Header component for improved accessibility.
- Implemented search query handling and routing to the search results page.

tomast1337 committed:
Merge branch 'feature/song-search' of github.com:OpenNBS/NoteBlockWorld into feature/song-search

Bentroen commented on apps/backend/src/song/song.controller.ts:
```diff
@@ -67,147 +70,128 @@ export class SongController {
+        query.page,
+        query.limit,
+      );
+      return new PageDto<SongPreviewDto>({
```
Could this part of the code be extracted outside of the `if` clause? It's repeated exactly for each query mode :)

Bentroen commented on apps/frontend/src/app/(content)/page.tsx:
```diff
@@ -28,9 +25,8 @@ async function fetchRecentSongs() {
   try {
     const response = await axiosInstance.get<FeaturedSongsDto>(
-      '/song-browser/featured',
+      '/song/featured',
```
Isn't this route supposed to clash with the `GET /song/:id` endpoint? Is this currently the case, or is there some precedence order going on to prevent this from happening?

Our song IDs are 10 characters long, spanning `A-Za-z0-9`, but were they 8 characters long, `featured` could technically be a valid song ID.

Bentroen commented on apps/frontend/src/modules/browse/components/client/context/RecentSongs.context.tsx:
```diff
@@ -91,13 +96,13 @@ export function RecentSongsProvider({
     try {
       const response = await axiosInstance.get<Record<string, number>>(
-        '/song-browser/categories',
+        '/song/categories',
```
oop, there we go - `categories`, a perfectly valid song id! xD

Though I believe the chance of this happening is quite unlikely... 1 in 62^10 or something?

Bentroen reviewed (commented):
Awesome work! I'm so excited to ship this feature 🤩

I've left a few considerations in the review comments; not all of them may be actionable atm.

A few points we've yet to work on:
- I'll work on touching up the frontend presentation so it's more consistent with our current design.
- Recent songs seem not to be loading locally for me
- Are the tests still supposed to be failing?

Bentroen commented:
Pictures of the current frontend (courtesy of @tomast1337!):

<img width="600" height="993" alt="image" src="https://github.com/user-attachments/assets/ddf7c15d-b967-4ccc-8c19-8e5a1c53d035" />
<img width="600" height="993" alt="image" src="https://github.com/user-attachments/assets/0004e775-edf8-43ad-9dab-fae611fe1cc8" />

Bentroen committed:
Merge branch 'develop' into feature/song-search

Bentroen committed:
refactor: rename `SearchButton` to `SearchBar`

Bentroen committed:
fix: move search bar outside of popup and detach from button group

Bentroen committed:
feat: rework appearance of search bar

Bentroen committed:
fix: disable search bar auto-focus

Bentroen committed:
fix: don't clear search query when performing search

Bentroen committed:
feat: remove text logo, reorder NBW icon in navbar

Bentroen committed:
fix: rename `search-song` route to `search`

tomast1337 committed:
feat: integrate Typesense for enhanced song indexing and search capabilities

- Added Typesense service for indexing and searching songs.
- Implemented automatic indexing of unindexed songs using a scheduled job.
- Updated song controller to handle search queries via Typesense.
- Enhanced environment variables for Typesense configuration.
- Modified song entity to include a search indexing flag.
- Updated Docker Compose to include Typesense and its dashboard services.

tomast1337 committed:
Merge branch 'feature/song-search' of github.com:OpenNBS/NoteBlockWorld into feature/song-search

tomast1337 committed:
Revert "feat: integrate Typesense for enhanced song indexing and search capabilities"

This reverts commit 4fcf74f0fa45908b81e9ece2a42c38362508aac6.

tomast1337 committed:
chore: update dev script to use concurrently for running backend and frontend

tomast1337 committed:
chore: add @nbw/config as a workspace dependency in bun.lock and package.json

tomast1337 committed:
feat: enhance FeaturedSongsProvider to determine initial timespan based on available songs

- Added a helper function to find the first non-empty timespan or default to 'week'.
- Updated state initialization to use the determined initial timespan for featured songs.

tomast1337 committed:
feat: enhance song retrieval by adding category filtering for recent songs

- Updated SongController to support fetching recent songs by category.
- Modified RecentSongsProvider to handle new response structure with PageDto.
- Adjusted endpoint and data handling to accommodate category-based song retrieval.

tomast1337 committed:
chore: add zustand as a dependency in bun.lock and package.json

tomast1337 committed:
feat: implement song search with filters and state management using Zustand

- Introduced Zustand for managing song search state, including loading, pagination, and filter changes.
- Added filter options for category, sort, order, and results per page in the search interface.
- Enhanced the search functionality to update the URL based on selected filters and maintain state across searches.
- Implemented loading indicators and improved user experience with dynamic filtering and loading states.

tomast1337 committed:
feat: refactor search page components for improved structure and readability

- Introduced new components for the search page, including SearchPageSkeleton, LoadingOverlay, SearchHeader, SearchFilters, SearchResults, and NoResults.
- Enhanced the loading and no results states with dedicated components for better maintainability.
- Updated the main SearchSongPage component to utilize the new components, improving code organization and readability.

tomast1337 committed:
Merge branch 'develop' of github.com:OpenNBS/NoteBlockWorld into feature/song-search

tomast1337 committed:
chore(deps): update dependencies and configuration files

- Updated various dependencies in backend and configs, including:
  - Bumped AWS SDK packages to version 3.946.0
  - Upgraded NestJS packages to version 11.x
  - Updated class-validator, mongoose, and other related packages
- Added tailwindcss as a dependency in configs
- Modified TypeScript configuration files to use 'esnext' module and 'bundler' resolution
- Introduced new web-specific exports in the database package
